docs/source/design/input_output_specifications.md

```diff
@@ -0,0 +1,692 @@
+- **Asset management:** Facilitates consistent asset management by ensuring that all registered assets conform to expected input/output specifications, reducing manual errors and improving reproducibility.
+- **Improved user experience:** Users receive actionable feedback, making it easier to debug and resolve configuration issues prior to running benchmarks.
+
+
+
+### `InputOutputSpecs`
+
+
+```python
+class InputOutputSpecs(BaseModel, extra="forbid"):
+    name        : str                                 # Human-readable specification name
+    modality    : str                                 # e.g. "rna", "dna", "image"
+    semantic    : str                                 # Ontology reference (EDAM, OBI, GO, etc.)
+    organism    : str
+    version     : str = "1.0.0"                       # Increment when axes or encoding expectations change
+    axes        : Optional[List[Axis]] = None         # List of axes (e.g., cells, genes)
+    formats     : List[str] = ["h5ad"]                # Preferred file formats, ordered by preference
+    encodings   : Optional[List[str]] = ["dense"]     # Data encodings, e.g., "csr", "png"
+    validator   : Optional[str] = None                # Python function for validation (python:module:function)
+    validation_schema : Optional[Dict[str, Any]] = None # Optional schema for validator
+    description : Optional[str] = None                # Brief description
+
+    @property
+    def key(self) -> str:
+        """Canonical identifier used everywhere else."""
+        return f"{self.formats[0]}{self.modality}:{self.semantic}:{self.version}"
+```
+
+
+### `DatasetSpec`
+
+
+```python
+class DatasetSpec(BaseModel, extra="forbid"):
+    """
+    Describes a dataset's structure, provenance, and associated specifications.
+    """
+    name         : str                                 # Human-readable dataset name
+    description  : Optional[str] = None                # Brief description of the dataset
+    specs        : List[InputOutputSpecs]           # Unique set of input/output specifications
+    provenance   : Optional[Provenance] = None         # Dataset origin and authorship info
+    checksum     : Optional[str] = None                # Optional data integrity checksum
+    created_at   : Optional[datetime] = None           # Creation timestamp
+    metadata     : Dict[str, Any] = {}                 # Arbitrary additional metadata
+```
+
+
+###  `ModelSpec`
+
+
+```python
+class ModelSpec(BaseModel, extra="forbid"):
+    """
+    Describes a runnable model, its inputs, outputs, and metadata.
+    """
+    name              : str                                 # Human-readable model name
+    description       : Optional[str] = None                # Brief description of the model
+    version           : str                                 # Model version
+    required_inputs   : List[InputOutputSpecs] = []      # All required input specifications
```

Comment:
I don't see how this will map to the InputputOutputSpecs. What if a model requires anndata with certain metadata in the obs dataframe, and an external file of gene pathways (as an example). There may be many datasets with this required metadata in obs. And also one or more files that contain gene pathways, some of which may be acceptable and some of which aren't.

Can you explain what this would look like in practice?